docusaurus/docs/iOS/02-tutorials/02-audio-room-uikit.mdx

```diff
@@ -0,0 +1,1443 @@
+final class ParticipantsView: UIView {
+
+    private enum ParticipantsListSection: Int {
+        case speakers
+        case listeners
+    }
+
+    var content: [CallParticipant] = [] {
+        didSet { updateContent() }
+    }
+
+    private lazy var collectionViewDatasource: UICollectionViewDiffableDataSource<ParticipantsListSection, CallParticipant.ID> = makeDataSource()
+
+    private lazy var collectionView: UICollectionView = .init(frame: .zero, collectionViewLayout: makeCompositionalLayout())
+        .withoutTranslatesAutoresizingMaskIntoConstraints()
+
+    private let call: Call
+    private var cancellables: Set<AnyCancellable> = []
+
+    @available(*, unavailable)
+    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
+
+    init(call: Call) {
+        self.call = call
+
+        super.init(frame: .zero)
+
+        addSubview(collectionView)
+        collectionView.pin(to: self)
+
+        collectionView.dataSource = collectionViewDatasource
+        collectionView.register(
+            ParticipantCollectionViewCell.self,
+            forCellWithReuseIdentifier: "ParticipantCollectionViewCell"
+        )
+
+        call.state
+            .$participants
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] in self?.content = $0 }
+            .store(in: &cancellables)
+    }
+
+    func updateContent() {
+        guard !content.isEmpty else {
+            loadSnapshot(speakers: [], listeners: [])
+            return
+        }
+
+        var speakers: [CallParticipant] = []
+        var listeners: [CallParticipant] = []
+
+        content.forEach {
+            if $0.isSpeaking { speakers.append($0) }
+            else { listeners.append($0) }
+        }
+
+        loadSnapshot(speakers: speakers, listeners: listeners)
+    }
+
+    private func loadSnapshot(
+        speakers: [CallParticipant],
+        listeners: [CallParticipant]
+    ) {
+        var snapshot = NSDiffableDataSourceSnapshot<ParticipantsListSection, CallParticipant.ID>()
+        snapshot.appendSections([.speakers, .listeners])
+        snapshot.appendItems(speakers.map(\.id), toSection: .speakers)
+        snapshot.appendItems(listeners.map(\.id), toSection: .listeners)
+        collectionViewDatasource.applySnapshotUsingReloadData(snapshot)
+    }
+
+    private func makeDataSource() -> UICollectionViewDiffableDataSource<ParticipantsListSection, CallParticipant.ID> {
+        // Create a cell registration that the diffable data source will use.
+        let cellRegistration = UICollectionView.CellRegistration<ParticipantCollectionViewCell, CallParticipant> {
+            cell, indexPath, item in
+            cell.content = item
+        }
+
+        // Create the diffable data source and its cell provider.
+        return UICollectionViewDiffableDataSource(collectionView: collectionView) { [unowned self] collectionView, indexPath, identifier -> UICollectionViewCell in
```

Comment:
is it safe to use `unowned` here?

Reply:
It is as the datasource's lifecycle is bound to the view (and it can't leak). I had it as weak in the beginning but i needed to return an instance of UICollectionViewCell if weak self has been deallocated and thought that with unowned it's easier to read.